zmq_addon.hpp

```diff
@@ -408,190 +408,115 @@ class multipart_t
-
-        str = received.popstr();
-        assert(received.size() == 2);
-        assert(str == "Fifty");
-
-        str = received.popstr();
-        assert(received.size() == 1);
-        assert(str == "Ninety-nine");
-
-        str = received.popstr();
-        assert(received.empty());
-        assert(str == "One-hundred");
-
-        return 0;
-    }
-
 private:
     // Disable implicit copying (moving is more efficient)
     multipart_t(const multipart_t& other) ZMQ_DELETED_FUNCTION;
     void operator=(const multipart_t& other) ZMQ_DELETED_FUNCTION;
 };  // class multipart_t
 
-#endif  // ZMQ_HAS_RVALUE_REFS
-
 inline std::ostream& operator<<(std::ostream& os, const multipart_t& msg)
 {
     return os << msg.str();
 }
 
+#endif  // ZMQ_HAS_RVALUE_REFS
+
+#if defined(ZMQ_BUILD_DRAFT_API) && defined(ZMQ_CPP11) && defined(ZMQ_HAVE_POLLER)
+    class active_poller_t
+    {
+    public:
+        active_poller_t () = default;
+        ~active_poller_t () = default;
+
+        active_poller_t(const active_poller_t&) = delete;
+        active_poller_t &operator=(const active_poller_t&) = delete;
+
+        active_poller_t(active_poller_t&& src) = default;
+        active_poller_t &operator=(active_poller_t&& src) = default;
+
+        using handler_t = std::function<void(short)>;
+
+        void add (zmq::socket_t &socket, short events, handler_t handler)
+        {
+            auto it = decltype (handlers)::iterator {};
+            auto inserted = bool {};
+            std::tie(it, inserted) = handlers.emplace (static_cast<void*>(socket), std::make_shared<handler_t> (std::move (handler)));
+            try 
+            {
+                base_poller.add (socket, events, inserted && *(it->second) ? it->second.get() : nullptr);
+                need_rebuild |= inserted;
+            }
+            catch (const zmq::error_t&) 
+            {
+                // rollback
+                if (inserted)
+                {
+                    handlers.erase (static_cast<void*>(socket));
+                }
+                throw;
+            }
+        }
+
+        void remove (zmq::socket_t &socket)
+        {
+            base_poller.remove (socket);
+            handlers.erase (static_cast<void*>(socket));
+            need_rebuild = true;
+        }
+
+        void modify (zmq::socket_t &socket, short events)
+        {
+            base_poller.modify (socket, events);
+        }
+
+        int wait (std::chrono::milliseconds timeout)
+        {
+            if (need_rebuild) {
+                poller_events.resize (handlers.size ());
+                poller_handlers.clear ();
+                poller_handlers.reserve (handlers.size ());
+                for (const auto &handler : handlers) {
+                    poller_handlers.push_back (handler.second);
+                }
+                need_rebuild = false;
+            }
+            const int count = base_poller.wait_all (poller_events, timeout);
+            if (count != 0) {
```

Comment:
count > 0

Reply:
Yes, I agree, that would be better.

Reply:
Actually, the condition is completely redundant. The for_each below just iterates over an empty range if count == 0.